android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/TesterActivity.java

```diff
@@ -0,0 +1,141 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.android.integrationtest;
+
+import com.google.android.gms.security.ProviderInstaller;
+
+import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+import android.support.v7.app.AppCompatActivity;
+import android.text.TextUtils;
+import android.util.Log;
+import android.view.View;
+import android.view.inputmethod.InputMethodManager;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class TesterActivity extends AppCompatActivity
+    implements ProviderInstaller.ProviderInstallListener {
+  final static String LOG_TAG = "GrpcTest";
+  private List<Button> buttons;
+  private EditText hostEdit;
+  private EditText portEdit;
+  private TextView resultText;
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_tester);
+    buttons = new LinkedList<>();
+    buttons.add((Button) findViewById(R.id.empty_unary_button));
+    buttons.add((Button) findViewById(R.id.large_unary_button));
+    buttons.add((Button) findViewById(R.id.client_streaming_button));
+    buttons.add((Button) findViewById(R.id.server_streaming_button));
+    buttons.add((Button) findViewById(R.id.ping_pong_button));
+
+    hostEdit = (EditText) findViewById(R.id.host_edit_text);
+    portEdit = (EditText) findViewById(R.id.port_edit_text);
+    resultText = (TextView) findViewById(R.id.grpc_response_text);
+
+    ProviderInstaller.installIfNeededAsync(this, this);
+    // Disable buttons until the security provider installing finishes.
+    enableButtons(false);
+  }
+
+  public void startEmptyUnary(View view) {
+    startTest("empty_unary");
+  }
+
+  public void startLargeUnary(View view) {
+    startTest("large_unary");
+  }
+
+  public void startClientStreaming(View view) {
+    startTest("client_streaming");
+  }
+
+  public void startServerStreaming(View view) {
+    startTest("server_streaming");
+  }
+
+  public void startPingPong(View view) {
+    startTest("ping_pong");
+  }
+
+  private void enableButtons(boolean enable) {
+    for (Button button : buttons) {
+      button.setEnabled(enable);
+    }
+  }
+
+  private void startTest(String testCase) {
+    ((InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE)).hideSoftInputFromWindow(
+        hostEdit.getWindowToken(), 0);
+    enableButtons(false);
+    String host = hostEdit.getText().toString();
+    String portStr = portEdit.getText().toString();
+    int port = TextUtils.isEmpty(portStr) ? 0 : Integer.valueOf(portStr);
```

Comment:
Why use port 0? That is highly unlikely to be the right port... Same in TesterInstrumentation. I'd suggest (in preference order) 1) using port 8080 (as this is what the test server is by default), 2) using port 443, 3) throwing an exception, 4) using port 8030